portfolio/src/main/java/com/google/sps/servlets/AddCommentServlet.java

```diff
@@ -34,15 +45,52 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
     Date postDate = new Date();
     String text = request.getParameter("comment-text");
 
+    // Get the URL of the image that the user uploaded to Blobstore.
+    String imageUrl = getUploadedFileUrl(request, "image");
+
     // Create comment entity 
     Entity commentEntity = new Entity("BlogComment");
     commentEntity.setProperty("text", text);
     commentEntity.setProperty("postDate", postDate);
+    commentEntity.setProperty("imageURL", imageUrl);
 
     // Upload comment to Datastore 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
 
     response.sendRedirect("/blog.html");
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+    String url = imagesService.getServingUrl(options);
+
+    // GCS's localhost preview is not actually on localhost, so make the URL relative to the current domain.
+    if(url.startsWith("http://localhost:8080/")){
```

Comment:
Based on our talk of Clean code. This method does not just return image url but also changes url to take care of this side effect. So this method is doing more than one thing, and when writing test we may need to add extra cases to test this side effect

Reply:
I see your point, I will make a separate function to handle this URL formatting!